distributed/worker_state_machine.py

```diff
@@ -3666,19 +3666,18 @@ def handle_stimulus(self, *stims: StateMachineEvent) -> None:
                 self._async_instructions.add(task)
                 task.add_done_callback(self._handle_stimulus_from_task)
 
-    async def close(self, timeout: float = 30) -> None:
+    async def close(self) -> None:
         """Cancel all asynchronous instructions"""
         if not self._async_instructions:
             return
-        for task in self._async_instructions:
-            task.cancel()
-        # async tasks can handle cancellation and could take an arbitrary amount
-        # of time to terminate
-        _, pending = await asyncio.wait(self._async_instructions, timeout=timeout)
-        for task in pending:
-            logger.error(
-                f"Failed to cancel asyncio task after {timeout} seconds: {task}"
-            )
+        pending_to_cancel = self._async_instructions
+        while pending_to_cancel:
+            for task in self._async_instructions:
+                task.cancel()
+            # async tasks can handle cancellation and could take an arbitrary amount
+            # of time to terminate
+            _, pending = await asyncio.wait(self._async_instructions, timeout=0.1)
+            pending_to_cancel.update(pending)
```

Comment:
I don't really know how I feel about this. I think it's entirely unnecessary since we're not handling CancelledErrors. I think just cancelling is fine. There is one artificial test which provokes this but I'd be fine deleting that test.

@graingert curious to get your thoughts about this

Reply:
For context, the tasks that can be in async_instructions is *very* limited. It needs to belong to an [`Instruction`](https://github.com/dask/distributed/blob/69b74f9f0aa46dac7facb960f1b0aaac69121e5e/distributed/worker_state_machine.py#L354)

which is implemneted here 
https://github.com/dask/distributed/blob/69b74f9f0aa46dac7facb960f1b0aaac69121e5e/distributed/worker_state_machine.py#L3629-L3667

On a side note, I think we can/should just use our AsyncTaskGroup for this but I would prefer doing this in another PR

Reply:
1. I think our current implementation of async user functions is leaky in the sense that I don't want their coroutines be executed on our event loop. Instead, we should still dispatch this to a threadpool and in this threadpool they have their own event loop
2. With my changes nothing would be stuck but I'd abort their graceful shutdown by cancelling it as well.

> Worth noting however that this behaviour is incoherent with sync tasks, which now run forever.

They always wait forever. There is now way to cancel/abort/interrupt a running thread. See https://github.com/dask/distributed/issues/4694 for a discussion and https://github.com/dask/distributed/pull/4726 for a prototype

Reply:
> I think our current implementation of async user functions is leaky in the sense that I don't want their coroutines be executed on our event loop. Instead, we should still dispatch this to a threadpool and in this threadpool they have their own event loop

Yes, I had this exact problem in my previous job, where I built an application on top of dask. The application was CPU-heavy, interleaved async db calls. I started my own event loop in a thread and it worked quite well.